### PR TITLE
AO-14676 graphql instrumentation

### PR DESCRIPTION
**Pull request by prettycoder**
there is code and tests for graphql

there is a also a little bit of cleanup of strings
- '' instead of ""
- freeze strings 

and other cleanup

### CODE REVIEW

File: lib/appoptics_apm/inst/graphql.rb [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+              end
+            end
+          end
+
+          def maybe_set_transaction_name(query)
+            return if gql_config[:transaction_name] == false ||
+                      ::AppOpticsAPM::SDK.get_transaction_name
+
+            split_query = query.strip.split(/\W+/, 3)
+            name = "graphql.#{split_query[0..1].join('.')}"
```
Comment (molafson) [inline lib/appoptics_apm/inst/graphql.rb]: Would the values in the range [0..1] contain anything that would be removed by the query sanitize (in other words this wouldn't set part of the transaction name to something that query sanitize would be removing for the query field)?

File: lib/appoptics_apm/inst/graphql.rb [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                AppOpticsAPM::SDK.log_info(Message: "GraphQL Errors:\n#{msg}")
+              end
+            end
+          end
+
+          def maybe_set_transaction_name(query)
+            return if gql_config[:transaction_name] == false ||
+                      ::AppOpticsAPM::SDK.get_transaction_name
+
+            split_query = query.strip.split(/\W+/, 3)
```
Comment (molafson) [inline lib/appoptics_apm/inst/graphql.rb]: Can the part of the query being use for the transaction name contain characters that aren't A-Za-z (eg é) as \W+ won't handle them?

File: test/instrumentation/graphql_test.rb [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        AppOpticsTest::Schema.execute(query)
+
+        traces = get_all_traces
+        traces.each do |tr|
+          refute_match '#', tr[:InboundQuery] if tr[:InboundQuery]
+        end
+      end
+    end
+
+    it 'removes comments if remove_comment is FALSE' do
```
Comment (molafson) [inline test/instrumentation/graphql_test.rb]: should be: `it "does not remove...`

File: lib/appoptics_apm/inst/graphql.rb (hunk 2 above)
Comment (prettycoder) [reply lib/appoptics_apm/inst/graphql.rb]: All the non-word characters, e.g. '{' get filtered out this way and only word characters make it into the transaction name. Names in graphql are limited to /[_A-Za-z][_0-9A-Za-z]*/

File: test/instrumentation/graphql_test.rb (hunk 3 above)
Comment (prettycoder) [reply test/instrumentation/graphql_test.rb]: right!!

File: lib/appoptics_apm/inst/graphql.rb (hunk 2 above)
Comment (molafson) [reply lib/appoptics_apm/inst/graphql.rb]: Okay 👍

**Comment by molafson** [comment]
Looks good 👍

File: lib/appoptics_apm/inst/graphql.rb (hunk 1 above)
Comment (prettycoder) [reply lib/appoptics_apm/inst/graphql.rb]: It wouldn't, the first 2 words can't contain any parameters, they are basically 'query', 'mutation', or 'subscription' plus a type.

Comment (molafson) [reply lib/appoptics_apm/inst/graphql.rb]: okay, thanks for confirming.

**Pull request merged**